File: resource-gateway-test-kit/polyglot/typescript/outcome-observation-verifier.test.ts
import assert from "node:assert/strict";
import { readFile } from "node:fs/promises";
import { test } from "node:test";
import { fileURLToPath } from "node:url";

import { verifyFixture } from "./outcome-observation-verifier.ts";

const fixturePath = fileURLToPath(
  new URL(
    "../../../docs/schemas/resource-gateway-mirror/" +
      "authoritative-outcome-observation-stage1-v1.fixture.json",
    import.meta.url,
  ),
);

async function fixture(): Promise<Record<string, unknown>> {
  return JSON.parse(await readFile(fixturePath, "utf8")) as Record<
    string,
    unknown
  >;
}

function copy<T>(value: T): T {
  return structuredClone(value);
}

test("verifies the server-produced observation without Java dependencies", async () => {
  const result = verifyFixture(await fixture());

  assert.deepEqual(result, {
    outcome: "VERIFIED",
    reasonCode: "VERIFIED",
    observationId: "outcome-refund-boundary",
    observationFingerprint:
      "sha256:5a12fd6db51178921b8098a0e61c04cd0dfc33524c410a25cc9e77494574534d",
    unitId: "refund-boundary",
    reconciliation: "MATCH",
    keyId: "memory-ed25519:2598df82-59b9-4954-a5fd-ee08f7c50e89",
  });
});

test("rejects a producer claim that disagrees with the authority closure", async () => {
  const value = await fixture();
  const observation = copy(value.observation) as Record<string, unknown>;
  observation.reconciliation = "MISMATCH";
  value.observation = observation;

  assert.equal(
    verifyFixture(value).reasonCode,
    "OUTCOME_RECONCILIATION_DERIVATION_INVALID",
  );
});

test("rejects content-address, authority-identity, and signature tampering", async () => {
  const original = await fixture();

  const fingerprintTamper = copy(original);
  (
    fingerprintTamper.observation as Record<string, unknown>
  ).observationFingerprint =
    "sha256:ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff";
  assert.equal(
    verifyFixture(fingerprintTamper).reasonCode,
    "OUTCOME_OBSERVATION_FINGERPRINT_INVALID",
  );

  const authorityTamper = copy(original);
  const authorityObservation = authorityTamper.observation as Record<
    string,
    unknown
  >;
  const facts = authorityObservation.authorityFacts as Array<
    Record<string, unknown>
  >;
  facts[0].subjectFingerprint =
    "sha256:ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff";
  assert.equal(
    verifyFixture(authorityTamper).reasonCode,
    "OUTCOME_ATTRIBUTION_CLOSURE_INVALID",
  );

  const signatureTamper = copy(original);
  const signedObservation = signatureTamper.observation as Record<
    string,
    unknown
  >;
  const seal = signedObservation.observationSeal as Record<string, unknown>;
  const signature = String(seal.signature);
  seal.signature = `${signature.startsWith("A") ? "B" : "A"}${signature.slice(
    1,
  )}`;
  assert.equal(
    verifyFixture(signatureTamper).reasonCode,
    "OUTCOME_OBSERVATION_SIGNATURE_INVALID",
  );
});

test("rejects unknown fields instead of silently accepting protocol drift", async () => {
  const value = await fixture();
  (value.observation as Record<string, unknown>).producerSaysValid = true;

  assert.equal(
    verifyFixture(value).reasonCode,
    "OUTCOME_OBSERVATION_SCHEMA_INVALID",
  );
});

test("rejects a Java-noncanonical instant spelling", async () => {
  const value = await fixture();
  value.verificationTime = "2026-07-26T04:00:00.000Z";

  assert.equal(
    verifyFixture(value).reasonCode,
    "OUTCOME_OBSERVATION_VERIFICATION_TIME_INVALID",
  );
});

test("retired keys still verify historical evidence", async () => {
  const value = await fixture();
  (value.verificationKey as Record<string, unknown>).state = "RETIRED";

  assert.equal(verifyFixture(value).outcome, "VERIFIED");
});

File: resource-gateway-test-kit/polyglot/go/outcome_observation_verifier_test.go
package protocolcert

import (
	"bytes"
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"testing"
)

func TestVerifiesServerProducedObservationWithoutJavaDependencies(t *testing.T) {
	result := VerifyFixture(loadFixture(t))

	expected := VerificationResult{
		Outcome:                "VERIFIED",
		ReasonCode:             "VERIFIED",
		ObservationID:          "outcome-refund-boundary",
		ObservationFingerprint: "sha256:5a12fd6db51178921b8098a0e61c04cd0dfc33524c410a25cc9e77494574534d",
		UnitID:                 "refund-boundary",
		Reconciliation:         "MATCH",
		KeyID:                  "memory-ed25519:2598df82-59b9-4954-a5fd-ee08f7c50e89",
	}
	if result != expected {
		t.Fatalf("verification mismatch:\nactual:   %#v\nexpected: %#v", result, expected)
	}
}

func TestRejectsProducerClaimThatDisagreesWithAuthorityClosure(t *testing.T) {
	value := loadFixture(t)
	observation(value)["reconciliation"] = "MISMATCH"

	assertReason(t, value, "OUTCOME_RECONCILIATION_DERIVATION_INVALID")
}

func TestRejectsContentAddressAuthorityIdentityAndSignatureTampering(t *testing.T) {
	fingerprintTamper := loadFixture(t)
	observation(fingerprintTamper)["observationFingerprint"] =
		"sha256:ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff"
	assertReason(t, fingerprintTamper, "OUTCOME_OBSERVATION_FINGERPRINT_INVALID")

	authorityTamper := loadFixture(t)
	facts := observation(authorityTamper)["authorityFacts"].([]any)
	facts[0].(map[string]any)["subjectFingerprint"] =
		"sha256:ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff"
	assertReason(t, authorityTamper, "OUTCOME_ATTRIBUTION_CLOSURE_INVALID")

	signatureTamper := loadFixture(t)
	seal := observation(signatureTamper)["observationSeal"].(map[string]any)
	signature := seal["signature"].(string)
	replacement := "A"
	if signature[0] == 'A' {
		replacement = "B"
	}
	seal["signature"] = replacement + signature[1:]
	assertReason(t, signatureTamper, "OUTCOME_OBSERVATION_SIGNATURE_INVALID")
}

func TestRejectsUnknownFieldsInsteadOfAcceptingProtocolDrift(t *testing.T) {
	value := loadFixture(t)
	observation(value)["producerSaysValid"] = true

	assertReason(t, value, "OUTCOME_OBSERVATION_SCHEMA_INVALID")
}

func TestRejectsJavaNonCanonicalInstantSpelling(t *testing.T) {
	value := loadFixture(t)
	value.(map[string]any)["verificationTime"] = "2026-07-26T04:00:00.000Z"

	assertReason(t, value, "OUTCOME_OBSERVATION_VERIFICATION_TIME_INVALID")
}

func TestRetiredKeyStillVerifiesHistoricalEvidence(t *testing.T) {
	value := loadFixture(t)
	value.(map[string]any)["verificationKey"].(map[string]any)["state"] = "RETIRED"

	if result := VerifyFixture(value); result.Outcome != "VERIFIED" {
		t.Fatalf("retired historical key was rejected: %#v", result)
	}
}

func assertReason(t *testing.T, value any, expected string) {
	t.Helper()
	if actual := VerifyFixture(value).ReasonCode; actual != expected {
		t.Fatalf("reason mismatch: actual=%s expected=%s", actual, expected)
	}
}

func observation(value any) map[string]any {
	return value.(map[string]any)["observation"].(map[string]any)
}

func loadFixture(t *testing.T) any {
	t.Helper()
	_, source, _, ok := runtime.Caller(0)
	if !ok {
		t.Fatal("cannot locate test source")
	}
	path := filepath.Join(
		filepath.Dir(source),
		"..", "..", "..", "docs", "schemas", "resource-gateway-mirror",
		"authoritative-outcome-observation-stage1-v1.fixture.json",
	)
	content, err := os.ReadFile(path)
	if err != nil {
		t.Fatal(err)
	}
	decoder := json.NewDecoder(bytes.NewReader(content))
	decoder.UseNumber()
	var value any
	if err := decoder.Decode(&value); err != nil {
		t.Fatal(err)
	}
	return value
}
